[windows] Always pass fp128 arguments indirectly

tgross35 opened the pull request:
LLVM currently expects `__float128` to be both passed and returned in
xmm registers on Windows. However, this disagrees with the Windows
x86-64 calling convention [1], which indicates values larger than 64
bits should be passed indirectly.

Update LLVM's default Windows calling convention to pass `fp128`
directly.  Returning in xmm0 is unchanged since this seems like a
reasonable extrapolation of the ABI. With this patch, the calling
convention for `i128` and `f128` is the same.

GCC passes `__float128` indirectly, which this also matches. However, it
also returns indirectly, which is not done here. I intend to attempt a
GCC change to also return in `xmm0` rather than making that change here,
given the consistency with `i128`.

This corresponds to the frontend change in [2], see more details there.

[1]: https://learn.microsoft.com/en-us/cpp/build/x64-calling-convention?view=msvc-170
[2]: https://github.com/llvm/llvm-project/pull/115052

tgross35 committed:
[windows] Add fp128 ABI tests and Windows libcall tests (NFC)

Add Windows invocations to existing fp128 libcall tests and create a new
ABI test for x86.

tgross35 committed:
[windows] Always pass `fp128` arguments indirectly

LLVM currently expects `__float128` to be both passed and returned in
xmm registers on Windows. However, this disagrees with the Windows
x86-64 calling convention [1], which indicates values larger than 64
bits should be passed indirectly.

Update LLVM's default Windows calling convention to pass `fp128`
directly.  Returning in xmm0 is unchanged since this seems like a
reasonable extrapolation of the ABI. With this patch, the calling
convention for `i128` and `f128` is the same.

GCC passes `__float128` indirectly, which this also matches. However, it
also returns indirectly, which is not done here. I intend to attempt a
GCC change to also return in `xmm0` rather than making that change here,
given the consistency with `i128`.

This corresponds to the frontend change in [2], see more details there.

[1]: https://learn.microsoft.com/en-us/cpp/build/x64-calling-convention?view=msvc-170
[2]: https://github.com/llvm/llvm-project/pull/115052

tgross35 commented:
@rnk in https://github.com/llvm/llvm-project/pull/115052#issuecomment-2458268283 you mentioned that changes in `X86ISelLowering` may be needed - any idea what these would look like? The test changes here seem fine but I'm also not sure what failure modes I should be looking for.

rnk commented:
I went looking at `X86TargetLowering::LowerWin64_i128OP`. I think that logic exists to manually implement the calling convention logic that you've implemented here, and perhaps we just don't need this logic anymore. I can't think of any specific concerns.

tgross35 commented:
Thanks for looking into that. I think this should be all set then, along with https://github.com/llvm/llvm-project/pull/128848.

tgross35 commented:
Thanks for reviewing, I need somebody to land this for me.

First commit for both PRs is NFC

tgross35 commented:
@phoebewang would you mind getting this one too?

The first commit is NFC and should land separate

phoebewang commented:
I see some problems here:

- The tile says it's for windows but description says for both. So the title is incorrect?
- I'd expect more tests affected by this change, e.g., https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/X86/fp128-cast.ll#L1059 should store into memory and pass to __gttf2, right?
- ABI change needs to be documented in ReleaseNote

tgross35 commented:
> I see some problems here:
> 
>     * The tile says it's for windows but description says for both. So the title is incorrect?

This is Windows-only. Were you referring to "Update LLVM's default calling convention" in the description? I meant this to refer to the default Windows CC used for libcalls, updated this to be more clear.

The sysv ABI actually specifies __float128 as SSE+SSEUP and LLVM already does the right thing there.

>     * I'd expect more tests affected by this change, e.g., https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/X86/fp128-cast.ll#L1059 should store into memory and pass to __gttf2, right?

That test only runs Linux, should I add the Windows targets? Some builtins calls are covered in `llvm/test/CodeGen/X86/fp128-libcalls.ll`, e.g. the call to `__addtf3` here https://github.com/llvm/llvm-project/pull/128848/commits/e0a3738d82436a5f4fac59d8fa52737118bf4165#diff-ea9428e87a77d0898c35fe6d9ff99f0ed9ba709ec232da43bf9fe0915894ce3cL159-R176

>     * ABI change needs to be documented in ReleaseNote

Done. Also rebased so this is on top of https://github.com/llvm/llvm-project/pull/115052.

phoebewang commented:
> That test only runs Linux, should I add the Windows targets? Some builtins calls are covered in `llvm/test/CodeGen/X86/fp128-libcalls.ll`, e.g. the call to `__addtf3` here [e0a3738#diff-ea9428e87a77d0898c35fe6d9ff99f0ed9ba709ec232da43bf9fe0915894ce3cL159-R176](https://github.com/llvm/llvm-project/commit/e0a3738d82436a5f4fac59d8fa52737118bf4165#diff-ea9428e87a77d0898c35fe6d9ff99f0ed9ba709ec232da43bf9fe0915894ce3cL159-R176)

No, I just read the patch as it affects both Linux and Windows. It looks good to me if it's Windows only.
